pkg/handler/collector/github/github.go

```diff
@@ -0,0 +1,206 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+)
+
+const (
+	CollectorGitHub = "GitHubCollector"
+)
+
+type githubCollector struct {
+	poll     bool
+	interval time.Duration
+	token    string
+	client   *github.Client
```

Comment:
In OSSF Scorecard, we realized that using the GitHub Client directly was making it difficult to test, so we refactored our code to use an interface and implementation instead. This will make the code more flexible and easier to test, as we can use a mock implementation of the interface in unit tests. Additionally, we replaced the GitHub response with a Custom Struct to make it compatible with other services, such as GitLab.


How about something like this instead of using the actual GitHub client?
```go
type RepoClient interface {
	GetLatestRelease(ctx context.Context, owner, repo string) (*github.RepositoryRelease, *github.Response, error)
	GetReleaseByTag(ctx context.Context, owner, repo, tag string) (*github.RepositoryRelease, *github.Response, error)
	GetCommitSHA1(ctx context.Context, owner, repo, ref, sha string) (string, *github.Response, error)
}

type GithubClient struct {
	client *github.Client
}

func (g *GithubClient) GetLatestRelease(ctx context.Context, owner, repo string) (*github.RepositoryRelease, *github.Response, error) {
	return g.client.Repositories.GetLatestRelease(ctx, owner, repo)
}

func (g *GithubClient) GetReleaseByTag(ctx context.Context, owner, repo, tag string) (*github.RepositoryRelease, *github.Response, error) {
	return g.client.Repositories.GetReleaseByTag(ctx, owner, repo,tag)
}

func (g *GithubClient) GetCommitSHA1(ctx context.Context, owner, repo, ref, sha string) (string, *github.Response, error) {
    return g.client.Repositories.GetCommitSHA1(ctx, owner, repo, ref, sha)
}